File: src/main/java/by/dev/madhead/jarvis/Jarvis.java
package by.dev.madhead.jarvis;

import by.dev.madhead.jarvis.model.Email;
import by.dev.madhead.jarvis.util.MessageBuilder;
import hudson.tasks.Mailer;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.Level;

import javax.mail.Address;
import javax.mail.Authenticator;
import javax.mail.MessagingException;
import javax.mail.PasswordAuthentication;
import javax.mail.Session;
import javax.mail.Transport;
import java.io.IOException;
import java.util.Properties;
import java.util.Set;

public class Jarvis {

    private static final Logger LOGGER = LogManager.getLogger();

    private final Session session;

    public Jarvis(boolean tlsEnable) {
        Mailer.DescriptorImpl mailerDescriptor = Mailer.descriptor();
        Properties properties = new Properties();
        properties.put("mail.smtp.host", mailerDescriptor.getSmtpServer());
        properties.put("mail.smtp.port", mailerDescriptor.getSmtpPort());
        properties.put("mail.smtp.auth", mailerDescriptor.getSmtpAuthUserName() != null);
        properties.put("mail.smtp.starttls.enable", tlsEnable);

        this.session = Session.getDefaultInstance(properties, mailerDescriptor.getSmtpAuthUserName() != null ?
                new Authenticator() {
                    @Override
                    protected PasswordAuthentication getPasswordAuthentication() {
                        return new PasswordAuthentication(mailerDescriptor.getSmtpAuthUserName(), mailerDescriptor.getSmtpAuthPassword());
                    }
                } : null);
    }

    public void sendMail(Email email, Set<Address> defaultRecipients) throws MessagingException, IOException {
        MessageBuilder messageBuilder = new MessageBuilder(email, defaultRecipients);
        if (messageBuilder.isMsgHasRecipients()) {
            Transport.send(messageBuilder.buildMessage(session));
        } else {
            LOGGER.log(Level.INFO, "No recipients to send mail.");
        }
    }

}

File: src/main/kotlin/by/dev/madhead/jarvis/Jarvis.kt
package by.dev.madhead.jarvis

import by.dev.madhead.jarvis.model.Email
import by.dev.madhead.jarvis.util.MessageBuilder
import hudson.AbortException
import hudson.tasks.Mailer
import org.apache.logging.log4j.Level
import org.apache.logging.log4j.LogManager
import java.io.IOException
import java.io.PrintStream
import java.util.*
import javax.mail.*

class Jarvis {

    private val session = fun(): Session {
        val mailerDescriptor = Mailer.descriptor()
        return Session.getDefaultInstance(
                Properties().apply {
                    this["mail.smtp.host"] = mailerDescriptor.smtpServer
                    this["mail.smtp.port"] = mailerDescriptor.smtpPort
                    this["mail.smtp.auth"] = (!mailerDescriptor.smtpAuthUserName.isNullOrBlank()).toString()
                    this["mail.smtp.ssl.enable"] = mailerDescriptor.useSsl
                },
                if (!mailerDescriptor.smtpAuthUserName.isNullOrBlank()) {
                    object : Authenticator() {
                        override fun getPasswordAuthentication(): PasswordAuthentication {
                            return PasswordAuthentication(
                                    mailerDescriptor.smtpAuthUserName,
                                    mailerDescriptor.smtpAuthPassword
                            )
                        }
                    }
                } else {
                    null
                }
        )
    }()

    fun sendMail(email: Email, defaultRecipientsAddresses: Set<Address>, logger: PrintStream) {
        val messageBuilder = MessageBuilder(email, defaultRecipientsAddresses)
        if (messageBuilder.isMsgHasRecipients()) {
            try {
                Transport.send(messageBuilder.buildMessage(session))
                logger.println("Emails sent to: ${messageBuilder.recipientsInfo()}")
            } catch (e: MessagingException) {
                handleException(logger, e)
            } catch (e: IOException) {
                handleException(logger, e)
            }
        } else {
            logger.println("No recipients to send email.")
        }
    }

    private fun handleException(jenkinsLogger: PrintStream, e: Exception) {
        jenkinsLogger.println("Error while sending email.")
        LogManager.getLogger().log(Level.ERROR, e.stackTrace)
        throw AbortException("${e.cause}: ${e.message}")
    }

}